### PR TITLE
feat: implement Chainlink relayer contract

### LINKED ISSUE #458: Implement Chainlink Relayer Contract
Implement relayer contract [according to spec](https://www.notion.so/mentolabs/v1-5-Chainlink-Relayer-Tech-Spec-230fd1ff82b14102a74c7edc5791f566?pvs=4#f16893bacea34a28afba0a310885a935)

- We have an open PR with a first draft already: https://github.com/mento-protocol/mento-core/pull/456
- But this needs to be reworked a bit after we made some design tweaks

### Requirements
- [ ] Update WIP PR #456 in accordance with latest design from spec

### PR DESCRIPTION
**Pull request by m-chrzan**
### Description

Adds a ChainlinkAdapter contract which can relay anwers from a Chainlink AggregatorV3 to SortedOracles.

It assumes it is the only reporter for the given rate feed in SortedOracles (e.g. `greater`/`lesserKey` are not needed).

Every rate feed that relies on this mechanism should have a separate ChainlinkAdapter deployed, with the appropriate parameters passed to the constructor. To save gas, the adapter is stateless, unupgradeable, and relies on immutable parameters set in the constructor.

Next steps:

* Integration test of the contract.
* Off-chain component that can ping this adapter contract.
* Deployment tooling/docs for setting up a rate feed with the ChainlinkAdapter.

### Tested

Unit tests.

### Related issues

- Fixes #458

### CODE REVIEW

File: contracts/interfaces/IChainlinkAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity >=0.5.13 <0.8.19;
+
+interface IChainlinkAdapter {
```
Comment (chapati23) [inline contracts/interfaces/IChainlinkAdapter.sol:4]:
```suggestion
interface IChainlinkRelayer {
```

File: contracts/interfaces/IChainlinkAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  function token() external returns (address);
```
Comment (chapati23) [inline contracts/interfaces/IChainlinkAdapter.sol:5]:
```suggestion
  function rateFeedId() external returns (address);
```

File: contracts/interfaces/IChainlinkAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+  function sortedOracles() external returns (address);
+
+  function aggregator() external returns (address);
```
Comment (chapati23) [inline contracts/interfaces/IChainlinkAdapter.sol:9]:
```suggestion
  function chainlinkAggregator() external returns (address);
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+pragma solidity ^0.8.18;
+
+import "../interfaces/IChainlinkAdapter.sol";
+import "foundry-chainlink-toolkit/src/interfaces/feeds/AggregatorV3Interface.sol";
+
+/**
+ * @notice The minimal subset of the SortedOracles interface needed by the
+ * adapter.
+ * @dev SortedOracles is a Solidity 5.17 contract, thus we can't import the
+ * interface directly, so we use a minimal hand-copied one.
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:11]: For convenience can we add a github link to SortedOracles interface here so it's easily verifiable that our hand-copying was accurate

File: contracts/oracles/ChainlinkAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    address greaterKey
+  ) external;
+
+  function medianTimestamp(address token) external view returns (uint256);
+
+  function getTokenReportExpirySeconds(address token) external view returns (uint256);
+}
+
+/**
+ * @title ChainlinkAdapter
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:27]:
```suggestion
 * @title ChainlinkRelayer
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+ * @notice The ChainlinkAdapter relays rate feed data from a Chainlink oracle to
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:28]:
```suggestion
 * @notice The ChainlinkRelayer relays rate feed data from a Chainlink price feed to
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+ * the SortedOracles contract. A separate instance should be deployed for each
+ * rate feed.
+ * @dev Assumes that it itself is the only reporter for the given SortedOracles
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:31]: hmm, to what extent is this assumption relevant here?

File: contracts/oracles/ChainlinkAdapter.sol [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+ * feed.
+ */
+contract ChainlinkAdapter is IChainlinkAdapter {
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:34]:
```suggestion
contract ChainlinkRelayer is IChainlinkRelayer {
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+  /**
+   * @notice The number of digits after the decimal point in FixidityLib
+   * values, as used by SortedOracles.
+   * @dev See contracts/common/FixidityLib.sol
+   */
+  uint256 public constant FIXIDITY_DECIMALS = 24;
+  /// @notice The rateFeedId this adapter relays for.
+  address public immutable token;
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:42]:
```suggestion
  address public immutable rateFeedId;
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+  /// @notice The address of the SortedOracles contract to report to.
+  address public immutable sortedOracles;
+  /**
+   * @notice The address of the Chainlink aggregator this contract fetches
+   * data from.
+   */
+  address public immutable aggregator;
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:49]:
```suggestion
  address public immutable chainlinkAggregator;
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+  /**
+   * @notice Used when a new answer's timestamp is not newer than the most recent
+   * SortedOracles timestamp.
+   */
+  error OldTimestamp();
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:55]: the more accurate name would be `NotANewTimestamp` or `NotNewerThan..` because `Old` suggests it is definitively older than SortedOracle's current one, when in reality there'll be many cases where the timestamps are identical between chainlink and sortedoracles, no?

struggling to find a succinct name that expresses this, some thoughts
- `TimestampNotNew()`
- `TimestampNotNewerThanSortedOracles()`

key point being that "old" is subtly different from "not new" in our case I think

File: contracts/oracles/ChainlinkAdapter.sol [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+  /// @notice The address of the SortedOracles contract to report to.
+  address public immutable sortedOracles;
+  /**
+   * @notice The address of the Chainlink aggregator this contract fetches
+   * data from.
+   */
+  address public immutable aggregator;
+
+  /**
+   * @notice Used when a new answer's timestamp is not newer than the most recent
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:52]:
```suggestion
   * @notice Used when a new price's timestamp is not newer than the most recent
```

I know chainlink calls this `answer` in their code, but feel like `price` is the more intuitive and concrete word here

File: contracts/oracles/ChainlinkAdapter.sol [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+   * SortedOracles timestamp.
+   */
+  error OldTimestamp();
+  /**
+   * @notice Used when a new answer's timestamp would be considered expired by
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:57]:
```suggestion
   * @notice Used when a new price's timestamp would be considered expired by
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+   * SortedOracles.
+   */
+  error ExpiredTimestamp();
+  /**
+   * @notice Used when a negative answer is returned by the Chainlink
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:62]:
```suggestion
   * @notice Used when a negative price is returned by the Chainlink
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+   * aggregator.
+   */
+  error NegativeAnswer();
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:65]:
```suggestion
  error NegativePrice();
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+  /**
+   * @notice Initializes the contract, setting immutable parameters.
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:68]:
```suggestion
   * @notice Initializes the contract and sets immutable parameters.
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+   * @param _token Address of the rate feed this adapter relays for.
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:69]:
```suggestion
   * @param _rateFeedId ID of the rate feed this relayer instance relays for.
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+   * @param _sortedOracles Address of the SortedOracles contract to relay to.
+   * @param _aggregator Address of the Chainlink aggregator to fetch data from.
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:71]:
```suggestion
   * @param _chainlinkAggregator Address of the Chainlink price feed to fetch data from.
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+   */
+  constructor(
+    address _token,
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:74]:
```suggestion
    address _rateFeedId,
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+    address _sortedOracles,
+    address _aggregator
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:76]:
```suggestion
    address _chainlinkAggregator
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+  ) {
+    token = _token;
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:78]:
```suggestion
    rateFeedId = _rateFeedId;
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+    sortedOracles = _sortedOracles;
+    aggregator = _aggregator;
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:80]:
```suggestion
    chainlinkAggregator = _chainlinkAggregator;
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  /**
+   * @notice Relays data from the configured Chainlink aggregator to
+   * SortedOracles.
+   * @dev Checks the answer is non-negative (Chainlink uses `int256` rather
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:86]:
```suggestion
   * @dev Checks the price is non-negative (Chainlink uses `int256` rather
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+   * than `uint256`.
+   * @dev Converts the answer to a Fixidity value, as expected by
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:88]:
```suggestion
   * @dev Converts the price to a Fixidity value, as expected by
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+   * SortedOracles.
+   * @dev Performs checks on the timestamp, will revert if any fails:
+   *      - The timestamp should be strictly newer than the most recent
+   *      timestamp in SortedOracles.
+   *      - The timestamp should not be considered expired by SortedOracles.
+   */
+  function relay() external {
+    ISortedOraclesMin _sortedOracles = ISortedOraclesMin(sortedOracles);
+    (, int256 answer, , uint256 timestamp, ) = AggregatorV3Interface(aggregator).latestRoundData();
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:97]:
```suggestion
    (, int256 answer, , uint256 timestamp, ) = AggregatorV3Interface(chainlinkAggregator).latestRoundData();
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+
+    uint256 lastTimestamp = _sortedOracles.medianTimestamp(token);
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:99]:
```suggestion
    uint256 lastTimestamp = _sortedOracles.medianTimestamp(rateFeedId);
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+        revert OldTimestamp();
+      }
+
+      if (isTimestampExpired(timestamp)) {
+        revert ExpiredTimestamp();
+      }
+    }
+
+    if (answer < 0) {
+      revert NegativeAnswer();
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:112]:
```suggestion
      revert NegativePrice();
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 28]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    uint256 report = chainlinkToFixidity(answer);
+
+    ISortedOraclesMin(sortedOracles).report(token, report, address(0), address(0));
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:117]:
```suggestion
    ISortedOraclesMin(sortedOracles).report(rateFeedId, report, address(0), address(0));
```

File: contracts/oracles/ChainlinkAdapter.sol [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+        revert ExpiredTimestamp();
+      }
+    }
+
+    if (answer < 0) {
+      revert NegativeAnswer();
+    }
+
+    uint256 report = chainlinkToFixidity(answer);
+
```
Comment (chapati23) [inline contracts/oracles/ChainlinkAdapter.sol:116]: a comment on why `address(0)` is fine here would be nice to have imo

File: test/oracles/ChainlinkAdapter.t.sol [hunk 30]
```diff
@@ -0,0 +1,96 @@
...
+    assertEq(_sortedOracles, address(sortedOracles));
+  }
+
+  function test_constructorSetsAggregator() public {
+    address _aggregator = adapter.aggregator();
+    assertEq(_aggregator, address(aggregator));
+  }
+}
+
+contract ChainlinkAdapterTest_relay is ChainlinkAdapterTest {
```
Comment (chapati23) [inline test/oracles/ChainlinkAdapter.t.sol:51]: think we should also explicitly test the correct conversion of `int256 answer` to `uint256 priceInFixidityFormat`

maybe also some fuzz tests to ensure a variety of rates all get converted correctly

**Review by chapati23 — CHANGES_REQUESTED** [review]

File: contracts/oracles/ChainlinkAdapter.sol (hunk 7 above)
Comment (m-chrzan) [reply contracts/oracles/ChainlinkAdapter.sol:31]: Just so future devs know this isn't a general Chainlink -> SortedOracles relayer that will work in every case, but instead needs to be in a setup where there aren't othe reporters for that rate feed (due to hardcoding lesser/greater keys to `address(0)`; technically, as discussed previously, this actually allows for up to *two* reporters for the feed, but no more).

File: test/oracles/ChainlinkAdapter.t.sol (hunk 30 above)
Comment (m-chrzan) [reply test/oracles/ChainlinkAdapter.t.sol]: Added [`testFuzz_convertsChainlinkToFixidityCorrectly`](https://github.com/mento-protocol/mento-core/pull/456/files#diff-e41049d6a1c786e606c257d3be3417c4f7e017e381f49232cfa326709f087353R66).

File: contracts/oracles/ChainlinkRelayer.sol [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+
+/**
+ * @title ChainlinkRelayer
+ * @notice The ChainlinkRelayer relays rate feed data from a Chainlink price feed to
+ * the SortedOracles contract. A separate instance should be deployed for each
+ * rate feed.
+ * @dev Assumes that it itself is the only reporter for the given SortedOracles
+ * feed.
+ */
+contract ChainlinkRelayer is IChainlinkRelayer {
```
Comment (chapati23) [inline contracts/oracles/ChainlinkRelayer.sol:35]: should we add the version here as well?

```suggestion
contract ChainlinkRelayerV1 is IChainlinkRelayer {
```

**Review by chapati23 — COMMENTED** [review]
looks great now, just one last question on version number

**Review by baroooo — CHANGES_REQUESTED** [review]

File: contracts/oracles/ChainlinkRelayerV1.sol [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+    (, int256 price, , uint256 timestamp, ) = AggregatorV3Interface(chainlinkAggregator).latestRoundData();
+
+    uint256 lastTimestamp = _sortedOracles.medianTimestamp(rateFeedId);
+
+    if (lastTimestamp > 0) {
+      if (timestamp <= lastTimestamp) {
+        revert TimestampNotNew();
+      }
+
+      if (isTimestampExpired(timestamp)) {
```
[the remark this answers is not in the document]
Comment (m-chrzan) [reply contracts/oracles/ChainlinkRelayerV1.sol]: Good catch. I don't think it would reject subsequent fresh rates, but yeah, could accept a stale rate in the edge case of the first report.

Fixed and added a regression test.

[the remark this answers is not in the document]
Comment (baroooo) [reply contracts/oracles/ChainlinkRelayerV1.sol]: yeah, I meant the subsequent "stale" rates would be rejected.
fix looks good!

**Review by baroooo — APPROVED** [review]
